File: backend/src/main/java/com/pm/backend/model/Task.java
package com.pm.backend.model;

import com.fasterxml.jackson.annotation.JsonIgnore;
import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
import jakarta.persistence.*;
import jakarta.validation.constraints.NotBlank;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;
import org.hibernate.annotations.CreationTimestamp;
import org.hibernate.annotations.UpdateTimestamp;

import java.time.LocalDate;
import java.time.LocalDateTime;
import java.util.HashSet;
import java.util.Set;

@Entity
@Table(name = "tasks")
@Data
@Builder
@NoArgsConstructor
@AllArgsConstructor
public class Task {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @NotBlank
    @Column(nullable = false)
    private String title;

    @Column(columnDefinition = "TEXT")
    private String description;

    @NotBlank
    @Column(nullable = false)
    private String status; // BACKLOG, TO_DO, IN_PROGRESS, TESTING, REVIEW, COMPLETED

    @NotBlank
    @Column(nullable = false)
    private String priority; // LOW, MEDIUM, HIGH, CRITICAL

    private LocalDate dueDate;

    @Builder.Default
    private Double estimatedTime = 0.0; // In hours
    @Builder.Default
    private Double actualTime = 0.0; // In hours

    @ManyToOne(fetch = FetchType.EAGER)
    @JoinColumn(name = "project_id", nullable = false)
    private Project project;

    @ManyToOne(fetch = FetchType.EAGER)
    @JoinColumn(name = "assignee_id")
    private User assignee;

    @ManyToOne(fetch = FetchType.EAGER)
    @JoinColumn(name = "reviewer_id")
    private User reviewer;

    @ManyToOne(fetch = FetchType.EAGER)
    @JoinColumn(name = "creator_id", nullable = false)
    private User creator;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "parent_id")
    @JsonIgnore
    private Task parentTask;

    @OneToMany(mappedBy = "parentTask", cascade = CascadeType.ALL, fetch = FetchType.LAZY)
    @Builder.Default
    private Set<Task> subtasks = new HashSet<>();

    @Builder.Default
    private Boolean recurring = false;
    private String recurringPattern; // DAILY, WEEKLY, MONTHLY

    @ManyToMany(fetch = FetchType.LAZY)
    @JoinTable(
        name = "task_dependencies",
        joinColumns = @JoinColumn(name = "task_id"),
        inverseJoinColumns = @JoinColumn(name = "dependency_id")
    )
    @JsonIgnoreProperties("dependencies")
    @Builder.Default
    private Set<Task> dependencies = new HashSet<>();

    @CreationTimestamp
    @Column(updatable = false)
    private LocalDateTime createdAt;

    @UpdateTimestamp
    private LocalDateTime updatedAt;
}

File: backend-node/src/models/Task.ts
import { DataTypes, Model, Optional } from 'sequelize';
import { sequelize } from '../config/database';

export interface TaskAttributes {
  id: number;
  title: string;
  description?: string;
  status: string;
  priority: string;
  dueDate?: string;
  estimatedTime?: number;
  actualTime?: number;
  projectId: number;
  assigneeId?: number | null;
  reviewerId?: number | null;
  creatorId: number;
  parentTaskId?: number | null;
  recurring?: boolean;
  recurringPattern?: string;
  declineReason?: string;
  acceptedAt?: Date | null;
  createdAt?: Date;
  updatedAt?: Date;
}

export interface TaskCreationAttributes extends Optional<TaskAttributes, 'id'> {}

export class Task extends Model<TaskAttributes, TaskCreationAttributes> implements TaskAttributes {
  declare id: number;
  declare title: string;
  declare description: string;
  declare status: string;
  declare priority: string;
  declare dueDate: string;
  declare estimatedTime: number;
  declare actualTime: number;
  declare projectId: number;
  declare assigneeId: number | null;
  declare reviewerId: number | null;
  declare creatorId: number;
  declare parentTaskId: number | null;
  declare recurring: boolean;
  declare recurringPattern: string;
  declare declineReason: string;
  declare acceptedAt: Date | null;

  declare readonly createdAt: Date;
  declare readonly updatedAt: Date;
}

Task.init(
  {
    id: {
      type: DataTypes.INTEGER,
      autoIncrement: true,
      primaryKey: true,
    },
    title: {
      type: DataTypes.STRING,
      allowNull: false,
    },
    description: DataTypes.TEXT,
    status: {
      type: DataTypes.STRING,
      allowNull: false,
      defaultValue: 'TO_DO',
    },
    priority: {
      type: DataTypes.STRING,
      allowNull: false,
      defaultValue: 'MEDIUM',
    },
    dueDate: DataTypes.STRING,
    estimatedTime: {
      type: DataTypes.FLOAT,
      defaultValue: 0.0,
    },
    actualTime: {
      type: DataTypes.FLOAT,
      defaultValue: 0.0,
    },
    projectId: {
      type: DataTypes.INTEGER,
      allowNull: false,
    },
    assigneeId: DataTypes.INTEGER,
    reviewerId: DataTypes.INTEGER,
    creatorId: {
      type: DataTypes.INTEGER,
      allowNull: false,
    },
    parentTaskId: DataTypes.INTEGER,
    recurring: {
      type: DataTypes.BOOLEAN,
      defaultValue: false,
    },
    recurringPattern: DataTypes.STRING,
    declineReason: DataTypes.TEXT,
    acceptedAt: DataTypes.DATE,
  },
  {
    sequelize,
    tableName: 'tasks',
  }
);
